### PR TITLE
ECM-38: Setup-Build-and-Dev-Tools-for-Frontend

### PR DESCRIPTION
**Pull request by larsduelfer**
Please test the build with the following two commands:
- npm install
- gulp build

Please try locally the following dev-servers (not in parallel):
- gulp serve-dev
- gulp serve-dist

### CODE REVIEW

File: frontend/src/app/app.component.html.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+          <li><a routerLink="/" routerLinkActive="active">Home</a></li>
+          <li><a id="casePage" routerLink="/cases" routerLinkActive="active">Cases</a></li>
+        </ul>
+        <ul class="nav navbar-nav navbar-right">
+          <!-- Insert the Login / Logout here -->
+        </ul>
+      </div>
+    </nav>
+    <router-outlet></router-outlet>
+`;
```
Comment (eberhardheber) [inline frontend/src/app/app.component.html.ts]: Please add a Line Break.

File: frontend/resources/styles/styles.scss [hunk 2]
```diff
@@ -20,8 +22,18 @@ body {
...
+
+.loading-spinner {
+  font-size: 3em;
+  color: $dark_blue;
+  position: fixed;
+  top: 50%;
+  left: 50%;
+  transform: translate(-50%, -50%);
+  text-align: center;
+}
```
Comment (eberhardheber) [inline frontend/resources/styles/styles.scss]: Please add a Line Break.

File: frontend/src/app/cases/cases.component.html.ts [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+    <control-panel-overview (loadCases)="loadCases($event)"></control-panel-overview>
+    <educama-tabset>
+        <educama-tab tabTitle="Instanzen / Transportauftrag">
+            <cases-instances></cases-instances>
+        </educama-tab>
+        <educama-tab tabTitle="Aufgaben">
+            <cases-tasks></cases-tasks>
+        </educama-tab>
+    </educama-tabset>
+`;
```
Comment (eberhardheber) [inline frontend/src/app/cases/cases.component.html.ts]: Please add a Line Break.

File: frontend/src/app/cases/cases_detail.component.html.ts [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+        <tr>
+            <th>Sendungsnummer:</th>
+            <td>{{ instance.id }}</td>
+        </tr>
+        <tr>
+            <th>Kunde:</th>
+            <td>{{ instance.customer }}</td>
+        </tr>
+    </table>
+`;
```
Comment (eberhardheber) [inline frontend/src/app/cases/cases_detail.component.html.ts]: Please add a Line Break.

File: frontend/src/app/cases/cases_list.component.html.ts [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+            <td>{{ case.id }}</td>
+            <td>{{ case.customer }}</td>
+            <td></td>
+            <td></td>
+        </tr>
+    </table>
+    <div *ngIf="selectedCase">
+        <button class="btn btn-primary" (click)="gotoDetail()">View Details</button>
+    </div>
+`;
```
Comment (eberhardheber) [inline frontend/src/app/cases/cases_list.component.html.ts]: Please add a Line Break.

File: frontend/src/app/cases/control/control_detail.component.html.ts [hunk 6]
```diff
@@ -0,0 +1,8 @@
+export const HTML_TEMPLATE = `
+    <div class="navbar navbar-default" role="navigation">
+        <h1 class="navbar-form pull-left" style="display: inline;">Auftrag {{ id }}</h1>
+        <form class="navbar-form pull-right">
+            <button class="btn btn-primary" (click)="back()">Zurück zur Übersicht</button>
+        </form>
+    </div>
+`;
```
Comment (eberhardheber) [inline frontend/src/app/cases/control/control_detail.component.html.ts]: Please add a Line Break.

File: frontend/src/app/cases/control/control_overview.component.html.ts [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+            <div class="navbar-header">
+                <a class="navbar-brand">Übersicht</a>
+            </div>
+            <form class="navbar-form pull-right">
+                <button id="addCase" class="btn btn-primary" (click)="createCaseInstance()">Neuen Transportauftrag anlegen</button>
+            </form>
+        </nav>
+        <create-instance-modal (modalOutput)="getCaseData($event)"></create-instance-modal>
+    </div>
+`;
```
Comment (eberhardheber) [inline frontend/src/app/cases/control/control_overview.component.html.ts]: Please add a Line Break.

File: frontend/src/app/cases/modal/create_instance_modal.component.html.ts [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+                    </div>
+                </div>
+                <div class="modal-footer">
+                    <button type="button" class="btn btn-default" (click)="close()">Cancel</button>
+                    <button id="newCase" type="button" class="btn btn-primary" (click)="submit(type.value, customer.value)">OK</button>
+                </div>
+            </div>
+        </div>
+    </div>
+`;
```
Comment (eberhardheber) [inline frontend/src/app/cases/modal/create_instance_modal.component.html.ts]: Please add a Line Break.

File: frontend/index.html [hunk 9]
```diff
@@ -5,29 +5,42 @@
...
+  <script src="node_modules/systemjs/dist/system.src.js"></script>
+  <script src="./systemjs.config.js"></script>
+  <script>
+    document.addEventListener('DOMContentLoaded', function () {
+      System.import('app')
+              .then(null, console.error.bind(console));
+    });
+  </script>
+  <!-- endbuild -->
+</body>
```
Comment (eberhardheber) [inline frontend/index.html]: Please add a Line Break.

File: frontend/src/app/cases/modal/create_instance_modal.component.html.ts [hunk 10]
```diff
@@ -0,0 +1,29 @@
+export const HTML_TEMPLATE = `
```
Comment (eberhardheber) [inline frontend/src/app/cases/modal/create_instance_modal.component.html.ts]: Why are you using Typscript files? It is possible to load only html files? I think it is better, because i cannot read it with an only html reader and i have to add typescript information like "export const HTML_TEMPLATE = `" and "`;" into the html text.

File: frontend/src/app/app.component.ts [hunk 11]
```diff
@@ -0,0 +1,11 @@
+import {Component} from "@angular/core";
+import {HTML_TEMPLATE} from "./app.component.html";
+
+@Component({
+    selector: 'educama-app',
```
Comment (eberhardheber) [inline frontend/src/app/app.component.ts]: You are mixing the signs ',` and". Can we use just one of these?

File: frontend/gulpfile.js [hunk 12]
```diff
@@ -0,0 +1,398 @@
...
+    startBrowsersync(browserSyncConfig.prod);
+});
+
+/**************************************************************
+ * Task definitions
+ *************************************************************/
+
+gulp.task('compile', ['sass', 'transpile']);
+gulp.task('build', ['typedoc']);
+gulp.task('test', ['remap-istanbul']);
```
Comment (eberhardheber) [inline frontend/gulpfile.js]: Please add a Line Break.

File: frontend/tslint.json [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+    ],
+    "component-selector-type": [
+      true,
+      "element"
+    ],
+    "directive-selector-prefix": [
+      true,
+      "ss"
+    ],
+    // should be activated again
```
Comment (eberhardheber) [inline frontend/tslint.json]: Do we need this comment? Or what is the problem with activating it?

File: frontend/tslint.json [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+    "component-class-suffix": true,
+    "directive-class-suffix": true,
+    "templates-use-public": true,
+    "no-access-missing-member": true,
+    "invoke-injectable": true,
+    "no-consecutive-blank-lines": true,
+    "no-unreachable": true,
+    "no-use-before-declare": true,
+    "no-var-keyword": true,
+    // should be activated again
```
Comment (eberhardheber) [inline frontend/tslint.json]: Do we need this comment? Or what is the problem with activating it?

Comment (larsduelfer) [inline frontend/tslint.json]: Yes, we would have a problem at the moment since we have not commited us yet to only use single or double quotes in the type script code and both is used right now. This can be changed with a separate pull request.

File: frontend/src/app/app.component.ts (hunk 11 above)
Comment (larsduelfer) [inline frontend/src/app/app.component.ts]: Yes we can and we should. See comment below. Once this is clarified, we need to change this in the entire app.

File: frontend/src/app/cases/modal/create_instance_modal.component.html.ts (hunk 10 above)
Comment (larsduelfer) [inline frontend/src/app/cases/modal/create_instance_modal.component.html.ts]: This is workaround to be able to bundle the app including the html templates for the components. The Angular CLI compiler will later be able to inline the external templates when compiling the components but for now this is the workaround.

**Comment by larsduelfer** [comment]
Hi Eberhard,

I commented some of your comments above. Regarding the line-breaks: Do you mean that every HTML tag should be in a separate line? I did not change the existing templates in the pull request. I only wrapped them in the type script constant. Please feel free to change this in a separate pull request.

Regards, Lars

File: frontend/gulpfile.js [hunk 15]
```diff
@@ -0,0 +1,398 @@
...
+gulp.task('watch', ['watch-ts', 'watch-sass']), function () {
+};
+
+/**************************************************************
+ * Serve App locally
+ *************************************************************/
+
+var browserSyncConfig = {
+    dev: {
+        port: 8000,
```
Comment (tobiasschaefer) [inline frontend/gulpfile.js]: Let's stay on port 8080.

File: frontend/gulpfile.js [hunk 16]
```diff
@@ -0,0 +1,398 @@
...
+            middleware: [historyApiFallback()]
+        },
+        files: [
+            buildDir + "/index.html",
+            buildDir + "/**/*.js",
+            resourceDir + "/**/*.css"
+        ]
+    },
+    prod: {
+        port: 8000,
```
Comment (tobiasschaefer) [inline frontend/gulpfile.js]: Let's stay on port 8080.

File: frontend/gulpfile.js [hunk 17]
```diff
@@ -0,0 +1,398 @@
...
+
+    asString: function () {
+        return process.env.VERSION || this.apiVersion
+            + '.' + this.productVersion
+            + '.' + this.buildNumber;
+    }
+};
+
+const targetName = 'educama-frontend';
+// targetFilename is used by deployLoacally and zip
```
Comment (tobiasschaefer) [inline frontend/gulpfile.js]: What is "deployLoacally"? (there is also a typo there)

File: frontend/gulpfile.js [hunk 18]
```diff
@@ -0,0 +1,398 @@
...
+    bsIns.reload();
+}
+
+/* Start live server dev mode */
+gulp.task('serve-dev', ['watch-ts', 'watch-sass'], function () {
+    startBrowsersync(browserSyncConfig.dev);
+});
+
+/* Start live server production mode */
+gulp.task('serve-dist', function () {
```
Comment (tobiasschaefer) [inline frontend/gulpfile.js]: Let's either use the term "dist" or "prod". Why mix them up?

File: frontend/index.html [hunk 19]
```diff
@@ -5,29 +5,42 @@
...
-  <link rel="stylesheet" href="styles.css">
+  <link data-require="bootstrap-css@*" data-semver="3.3.1" rel="stylesheet" href="//maxcdn.bootstrapcdn.com/bootstrap/3.3.1/css/bootstrap.min.css"/>
 
-  <!-- 1. Load libraries -->
-  <!-- Polyfill(s) for older browsers -->
-  <script src="https://unpkg.com/core-js/client/shim.min.js"></script>
+  <!-- build:css -->
+  <link rel="stylesheet" type="text/css" href="resources/primeng/font-awesome.min.css"/>
+  <link rel="stylesheet" type="text/css" href="/node_modules/primeng/resources/primeng.min.css"/>
+  <link rel="stylesheet" type="text/css"href="resources/styles/styles.css">
```
Comment (tobiasschaefer) [inline frontend/index.html]: missing closing xml tag here, please add a slash, i.e. ...styles.css"/>
